File: 1473-paint-house-iii/1473-paint-house-iii.py
class Solution:
    def minCost(self, houses: List[int], cost: List[List[int]], m: int, n: int, target: int) -> int:
        prev = [[inf] * (n + 1) for _ in range(target + 2)]
        curr = [[inf] * (n + 1) for _ in range(target + 2)]
        for i in reversed(range(m + 1)):
            for nhs in reversed(range(target + 1)):
                for pcolor in reversed(range(n + 1)):
                    if i == m:
                        curr[nhs][pcolor] = inf if nhs != target else 0
                        continue
                        
                    if houses[i] != 0:
                        curr[nhs][pcolor] = prev[
                            nhs + 1 if pcolor != houses[i] else nhs
                        ][houses[i]]
                        continue

                    for j in range(n):
                        curr[nhs][pcolor] = min(
                            curr[nhs][pcolor], 
                            cost[i][j] + prev[
                                nhs + 1 if pcolor != j + 1 else nhs
                            ][j + 1]
                        )
            prev, curr = curr, [[inf] * (n + 1) for _ in range(target + 2)]
        return prev[0][0] if prev[0][0] != inf else -1

File: 1473-paint-house-iii/1473-paint-house-iii.java
class Solution {
    public int minCost(int[] houses, int[][] cost, int m, int n, int target) {
        int[][][] dp = new int[m + 1][target + 1][n + 1];
        for(int i = 0; i <= m; i++)
            for(int j = 0; j <= target; j++)
                for(int k = 0; k <= n; k++)
                    dp[i][j][k] = -1;
        
        int res = helper(houses, cost, m, n, target, 0, 0, dp);
        return res == Integer.MAX_VALUE ? -1 : res;
    }
    
    private int helper(int[] houses, int[][] cost, int m, int n, int nhs, int i, int pcolor, int[][][] dp){
        if(i == m)
            return nhs == 0 ? 0 : Integer.MAX_VALUE;
        
        if(nhs < 0)
            return Integer.MAX_VALUE;
        
        if(dp[i][nhs][pcolor] != -1)
            return dp[i][nhs][pcolor];
        
        if(houses[i] != 0)
            dp[i][nhs][pcolor] = helper(
                houses, cost, m, n,
                pcolor != houses[i] ? nhs - 1 : nhs,
                i + 1, houses[i], dp
            );
        else{
            dp[i][nhs][pcolor] = Integer.MAX_VALUE;
            for(int j = 0; j < n; j++){
                int res = helper(
                    houses, cost, m, n,
                    pcolor != j + 1 ? nhs - 1 : nhs,
                    i + 1, j + 1, dp
                );

                if(res != Integer.MAX_VALUE)
                    dp[i][nhs][pcolor] = Math.min(
                        dp[i][nhs][pcolor],
                        cost[i][j] + res
                    );
            } 
        }
        return dp[i][nhs][pcolor];
    }
}